### PR TITLE
server: switch to options structure of upstream

### PR DESCRIPTION
**Pull request by sttts**
- use composable options+config structs from upstream (especially apiserver)
- with that make use of flag validation
- really test pkg/genericcontrollerplane in kcp-dev/k
- switch to `--run-controllers` enabling all controllers locally
- open up all the features from upstream components we will need for production anyway and don't want to rebuild
- add whitelist+blacklist of flags, secured against drift through unit tests
```
$ kcp start --help
Start the control plane process

The server process listens on port 6443 and will act like a Kubernetes API server. It will initialize any necessary data to the provided start location or as a '.kcp' directory in the current directory. An admin kubeconfig file will be generated at initialization time that may be used to access the control plane.


Usage:
  kcp start [flags]

Most important flags:

      --etcd-servers strings
                List of etcd servers to connect with (scheme://ip:port), comma separated. By default an embedded etcd
                server is started. (default [embedded])
      --run-controllers
                Run the controllers in-process (default true)

Use "kcp start options" for a list of all flags available.
```

https://docs.google.com/document/d/17SiAbgkj5o2lp9x9lzIqaSo8D65oP1sRz4mYCGNq1b4/edit with all the flags. Lots to be filtered out.

### CODE REVIEW

File: pkg/server/options/genericcontrolplane.go
Comment (stevekuznetsov) [inline pkg/server/options/genericcontrolplane.go:106]: ?

File: pkg/server/server.go
Comment (stevekuznetsov) [inline pkg/server/server.go:140]:
```suggestion
	// this needs surgery in k/k. So we do it here now, changing the already completed options. Not nice, but must be here because
```

**Review by stevekuznetsov — COMMENTED** [review]
Seems reasonable - is the `genericcontrolplane` change in k8s a follow-up?

File: pkg/server/options/embeddedetcd.go
Comment (smarterclayton) [inline pkg/server/options/embeddedetcd.go:44]: This name seems fine, although I'm a bit worried about adding flags that we don't directly and clearly need today (commented out or disabled here).  If WAL size is needed for integration I guess I can live with it.

File: pkg/server/options/genericcontrolplane.go
Comment (smarterclayton) [inline pkg/server/options/genericcontrolplane.go:40]: Would prefer fewer flags rather than more being exposed to users right now.  150+ flags is too many - I might prefer an allowlist mindset (if that hasn't already been thought of and i just missed it in review)

**Comment by smarterclayton** [comment]
To summarize, totally agree with the approach, but specifically the public api of number of flags has to be constrained and I would prefer only a very small allowlist of flags is exposed, and each one/set-of-similar-flags needs a use case justification

**Comment by sttts** [comment]
@smarterclayton please see https://docs.google.com/document/d/17SiAbgkj5o2lp9x9lzIqaSo8D65oP1sRz4mYCGNq1b4/edit and mark those you don't want to have. Maybe mark in some color.

File: pkg/server/options/genericcontrolplane.go
Comment (sttts) [reply pkg/server/options/genericcontrolplane.go:108]: legacyscheme is evil.

Comment (sttts) [reply pkg/server/options/genericcontrolplane.go:108]: The whole set of flags is now disabled.

File: pkg/server/options/genericcontrolplane.go
Comment (sttts) [reply pkg/server/options/genericcontrolplane.go:40]: agree, especially for these generic one. Most look like random one-offs.

File: pkg/server/options/embeddedetcd.go
Comment (sttts) [reply pkg/server/options/embeddedetcd.go]: Have removed it. Now `--etcd-servers=embedded` is the default.

```
Start the control plane process

The server process listens on port 6443 and will act like a Kubernetes API server. It will initialize any necessary data to the provided start location or as a '.kcp' directory in the current directory. An admin kubeconfig file will be generated at initialization time that may be used to access the control plane.


Usage:
  kcp start [flags]

Most important flags:

      --etcd-servers strings
                List of etcd servers to connect with (scheme://ip:port), comma separated. By default an embedded etcd
                server is started. (default [embedded])
      --run-controllers
                Run the controllers in-process (default true)

Use "kcp start options" for a list of all flags available.
```

**Comment by sttts** [comment]
> Seems reasonable - is the genericcontrolplane change in k8s a follow-up?

@stevekuznetsov yes, have an unfinished branch for that. Then I think the injector will be history 🥳

**Comment by sttts** [comment]
The failing tests seem to be timeouts. Locally I can run them without concurrency. Waiting for the CI PRs to land.

File: pkg/server/options/authorization.go
Comment (sttts) [inline pkg/server/options/authorization.go:92]: @s-urbaniak why was this not needed before? Without `/readyz` only works authenticated.

File: test/e2e/api_inheritance/api_inheritance_test.go
Comment (sttts) [inline test/e2e/api_inheritance/api_inheritance_test.go:85]: @ncdc how did this work before? The non-"admin" workspace never had the workspace CRD installed (and should not until we have a controller doing that for org workspaces).

File: pkg/server/options/authorization.go
Comment (s-urbaniak) [reply pkg/server/options/authorization.go:92]: it seems it was not wired before:
```
% curl -k https://127.0.0.1:6443/readyz
Internal Server Error: "/readyz": expected logical cluster name to be in org_name format, got admin
```

File: test/e2e/api_inheritance/api_inheritance_test.go
Comment (ncdc) [reply test/e2e/api_inheritance/api_inheritance_test.go:85]: This test used to only test Workspaces in admin.

**Review by davidfestal — APPROVED** [review]
Approving with just a very minimal typo suggestion.

File: cmd/kcp/help.go
Comment (sttts) [reply cmd/kcp/help.go:34]: odd that lint is not unhappy here.

File: pkg/server/options/flags_test.go
Comment (ncdc) [inline pkg/server/options/flags_test.go:36]:
```suggestion
				t.Errorf("--%s in section \"%s flags\" is neither allowed nor disallowed", f.Name, section)
```

**Review by ncdc — CHANGES_REQUESTED** [review]

**Review by ncdc — APPROVED** [review]

**Pull request merged**
